### PR TITLE
backupccl: elide spans from backups that were subsequently reintroduced

### PR DESCRIPTION
**Pull request by msbutler**
Currently RESTORE may restore invalid backup data from a backed up table that
underwent an IMPORT rollback. See https://github.com/cockroachdb/cockroach/issues/87305 for a detailed explanation.

This patch ensures that RESTORE elides older backup data that were deleted via
a non-MVCC operation. Because incremental backups always reintroduce spans
(i.e. backs them up from timestamp 0) that may have undergone a non-mvcc
operation, restore can identify restoring spans with potentially corrupt data
in the backup chain and only ingest the spans' reintroduced data to any system
time, without the corrupt data.

Here's the basic impliemenation in Restore:
- For each span we want to restore
   - identify the last time, l, the span was introduced, using the manifests
   - dont restore the span using a backup if backup.EndTime < l

This implementation rests on the following assumption: the input spans for each
restoration flow (created in createImportingDescriptors) and the
restoreSpanEntries (created by makeSimpleImportSpans) do not span across
multiple tables. Given this assumption, makeSimpleImportSpans skips adding
files from a backups for a given input span that was reintroduced in a
subsequent backup.

It's worth noting that all significant refactoring occurs on code run by
the restore coordinator; therefore, no special care needs to be taken for
mixed / cross version backups. In other words, if the coordinator has updated,
the cluster restores properly; else, the bug will exist on the restored cluster.
It's also worth noting that other forms of this bug are apparent on older
cluster versions (https://github.com/cockroachdb/cockroach/issues/88042, https://github.com/cockroachdb/cockroach/issues/88043) and has not been noticed by customers; thus,
there is no need to fail a mixed version restore to protect the customer from
this already existing bug.

Informs https://github.com/cockroachdb/cockroach/issues/87305

Release justification: bug fix

Release note: fix for TA advisory
https://cockroachlabs.atlassian.net/browse/TSE-198

### CODE REVIEW

**Comment by msbutler** [comment]
@dt @adityamaru only the last commit is relevant. I've verified this approach solves the data corruption problem. Does the basic approach look good? If so, I'll add a makeSimpleImportSpans unit test and add a few more comments.

**Comment by msbutler** [comment]
hmm it's breaking an AsOfSystemTime unit test. Let me figure out what's going on.

File: pkg/ccl/backupccl/restore_span_covering_test.go
Comment (msbutler) [inline pkg/ccl/backupccl/restore_span_covering_test.go:45]: I'm confused by a few parts of MockBackupChain: 
 - The `makeTableSpan(uint32(100 + j + (i / 3)))` call above implies that each backup[i].Span[j] is unique, as each gets created with a unique TableID (note that i is included in the TableID calc). Is that intended? I feel like the same spans should get covered over a series of backups.
 - In the line below, I randomized which span gets reintroduced. `TestRestoreEntryCover` continues to pass with this diff; however `BenchmarkCoverageChecks` now fails in the `checkCoverage` call. I don't understand why.

Comment (msbutler) [reply pkg/ccl/backupccl/restore_span_covering_test.go:45]: hrm. removing the `i` from the input call, seems to have solved the problem 🤷

File: pkg/ccl/backupccl/restore_job.go
Comment (adityamaru) [inline pkg/ccl/backupccl/restore_job.go:719]: nit: maybe rename to `latestEndTimesForIntroducedSpans`

File: pkg/ccl/backupccl/restoration_data.go
Comment (adityamaru) [inline pkg/ccl/backupccl/restoration_data.go:171]: why do we need this method, I think I'm missing something. Why can't we just operate on the introduced spans mapping that we created initially and in `makeSimpleImportSpans` just check if the span being processed is part of that mapping, and if it is then perform the relevant time filtering. I don't quite get why we map[{tableID,indexID}]:[]ts and then pivot that to []ts that is 1:1 with the passed in spans slice.

**Review by adityamaru — CHANGES_REQUESTED** [review]
Flushing out first round, but mainly want clarity on why we need `findLatestIntroBySpan` as explained below.

File: pkg/ccl/backupccl/restoration_data.go
Comment (msbutler) [reply pkg/ccl/backupccl/restoration_data.go:171]: good q. I initially implemented what you proposed, but chose the current approach for modularity. My high level thinking is: "during makeSimpleImportSpans, we operate with spans. we shouldn't get in the business of decoding keys for table IDs-- that work should happen before we call `restoreWithRetry`. 

On the other hand, maybe we should do what you're suggesting-- this function you highlight isn't strictly necessary. Further, we implicitly do operate on the index level in `makeSimpleImportSpans`, as each span represents an index. If you prefer your approach, I'm happy to change it.

File: pkg/ccl/backupccl/restore_job.go
Comment (msbutler) [reply pkg/ccl/backupccl/restore_job.go:668]: I think a lot of other backup and restore code would break if {tableID,IndexID} didn't 1:1 map to a unique backup or restore span. `roachpb.MergeSpans()` should never merge two spans from different indexes, right? We seem to rely on this invariant via the `tableAndIndex` struct throughout the backup and restore code.

Comment (adityamaru) [reply pkg/ccl/backupccl/restore_job.go:668]: >  should never merge two spans from different indexes

I think it merges contiguous spans even if they're from different indexes so [/table/12/1 - /table/12/2) and [/table/12/2 - /table/12/3) will merge to [/table/12/1-/table/12/3), and [/table/12/1-/table/12/2) + [/table/13/1-/table/13/2) merges to [/table/12/1 - /table/13/2)? That aside I do agree that we do rely on {tableID, indexID} elsewhere but is it correct to create a mapping based on `IntroducedSpans` created during backup and then elide **all** spans with the same {tableID, indexID} during restore. You might be right I just want to convince myself, so assume when taking an incremental backup we see table 12 comes online so `IntroducedSpans` will have [/table/12/1-/table/13/1) so when restoring we will run this check for spans where the start key is in table 12 and index 1 but what about table 12 and index 2?

Comment (msbutler) [reply pkg/ccl/backupccl/restore_job.go:668]: > I think it merges contiguous spans even if they're from different indexes so [/table/12/1 - /table/12/2) and [/table/12/2 - /table/12/3) will merge to [/table/12/1-/table/12/3), and [/table/12/1-/table/12/2) + [/table/13/1-/table/13/2) merges to [/table/12/1 - /table/13/2)

I don't think your examples are correct because the span for table 12, index 1, is [/table/12/1/0 - /table/12/1/{prefixEnd}), not [/table/12/1 - /table/12/2) -- i.e. the end key for a given index span will have the same prefix as the start key. Notice how index spans are created using a `PrefixEnd()` call within `forEachPublicIndexTableSpan` Given this, I don't _think_ spans from different indices overlap, and consequently merge.

Comment (adityamaru) [reply pkg/ccl/backupccl/restore_job.go:668]: Hmm you're right about the prefixEnd but I did the following:

```
CREATE TABLE foo (id INT PRIMARY KEY);
CREATE INDEX bar ON foo (id);
INSERT INTO foo VALUES (1);

root@localhost:26257/defaultdb> select crdb_internal.pretty_key(key, 0) from crdb_internal.scan(crdb_internal.table_span(104));
  crdb_internal.pretty_key
----------------------------
  /104/1/1/0
  /104/2/1/0
(2 rows)

root@localhost:26257/defaultdb> backup table foo into 'nodelocal://1/foo';
        job_id       |  status   | fraction_completed | rows | index_entries | bytes
---------------------+-----------+--------------------+------+---------------+--------
  795137589190295553 | succeeded |                  1 |    1 |             1 |    18
(1 row)
```

I added a print statement before and after `roachpb.MergeSpans()` in `spansForAllTableIndexes` and it shows me:
```
I220909 12:43:44.609424 2252 ccl/backupccl/backup_planning.go:174 ⋮ [n1,job=795138537598124033] 219  these are the unmerged spans [‹/Table/104/{1-2}› ‹/Table/104/{2-3}›]
I220909 12:43:44.609567 2252 ccl/backupccl/backup_planning.go:181 ⋮ [n1,job=795138537598124033] 220  these are the merged spans [‹/Table/104/{1-3}›]
```

Comment (msbutler) [reply pkg/ccl/backupccl/restore_job.go:668]: wow, great catch! It's quite subtle that the PrefixEnd (i.e. the endKey) of an index span can be the start key of a different index which causes index spans to merge; whereas that's not the case for table spans. I'll add a bit more documentation to `forEachPublicIndexTableSpan`, maintain a tableID map instead, and add some indices in my tests.

File: pkg/ccl/backupccl/restoration_data.go
Comment (adityamaru) [reply pkg/ccl/backupccl/restoration_data.go:171]: I think it simplifies things if we have a single method that creates the mapping and thats our source of truth throughout. We're not paying an additional cost in decoding the key since the number of spans we're passing into `makeSimpleImportSpans` is the same as if we did it individually.

File: pkg/ccl/backupccl/restore_job.go
Comment (msbutler) [reply pkg/ccl/backupccl/restore_job.go]: renamed to latestEndTimesForIntroducedTables

File: pkg/ccl/backupccl/restore_job.go
Comment (msbutler) [reply pkg/ccl/backupccl/restore_job.go]: deleted this function. no longer relevant.

**Comment by msbutler** [comment]
ah, I need to be more precise about which `IntroducedSpans` I want to use to use for filtering-- specifically, I must only consider spans that were reintroduced via non-mvcc shenanigans, and not say, a new index on a table. Back to the drawing board...

**Comment by msbutler** [comment]
Here's backup history example that the current, implementation cannot handle:

t0: create table foo, and add data to it
t1: run a full backup
t2: create an index foo_idx on foo
t3: run an incremental backup, adding foo_idx to backupManifest.IntroducedSpans (note that all of foo doesn't get backed up from t0, just the index foo_idx)
t4: restore foo
- because foo_idx is an introduced span, we naively add table foo to our map of `latestEndTimesForIntroducedTables` and don't include any data from the first backup in the restore. This leads to data loss.

The only way I think we can get around this is to include more data in the backup manifest. Specifically, we'll need to divide the `IntroducedSpans` field into `IntroducedSpans` (i.e. what foo_idx is) and a new field `ReIntroducedSpans` which contains all spans that need to be reintroduced due to non-mvcc shenanigans. Then, on the restore side, we can only do clean import rollbacks iff the whole backup chain was taken on a fully upgraded 22.2 cluster. I don't like that we need to manipulate the backup manifest, but I can't think of a better way that only touches restore.

File: pkg/ccl/backupccl/backupinfo/manifest_handling.go
Comment (adityamaru) [inline pkg/ccl/backupccl/backupinfo/manifest_handling.go:922]: I think this method works on the assumption that `manifests` is sorted in increasing EndTime because of how we are unconditionally overwriting the map value? If thats the case can we mention it in the comment.

File: pkg/ccl/backupccl/backupinfo/manifest_handling.go
Comment (adityamaru) [inline pkg/ccl/backupccl/backupinfo/manifest_handling.go:916]: nit: "endtime of the latest incremental backup that reintroduced each backed up table" confuses me a little because its not really "each backup up table", its only the tables that were reintroduced because of being offline at some point between backups.

Maybe we can say "finds the EndTime of the latest incremental backup for each table that was reintroduced because of being previously offline". See ReintroducedSpans() for more information.

File: pkg/ccl/backupccl/restore_span_covering.go
Comment (adityamaru) [inline pkg/ccl/backupccl/restore_span_covering.go:114]: I like this comment, but I'll like it even more with an example of the invalid data you're mentioning. Maybe copy paste it from the issue you've opened?

File: pkg/ccl/backupccl/targets.go
Comment (adityamaru) [inline pkg/ccl/backupccl/targets.go:495]: nit: why 1? why not just make(map[descpb.ID]hlc.Timestamp)

File: pkg/ccl/backupccl/bench_covering_test.go
Comment (adityamaru) [inline pkg/ccl/backupccl/bench_covering_test.go:61]: nit: can we make this the system tenant codec, we're trying to get rid of TODOSQLCodec so no point adding another?

File: pkg/ccl/backupccl/restore_span_covering_test.go
Comment (adityamaru) [inline pkg/ccl/backupccl/restore_span_covering_test.go:318]: can we use a system tenant codec?

File: pkg/ccl/backupccl/restore_span_covering_test.go
Comment (adityamaru) [inline pkg/ccl/backupccl/restore_span_covering_test.go:124]: nit: quick comment for future readers about why we're skipping these backups, even if you just point to the comment in makeSimpleImportSpans thats fine

File: pkg/ccl/backupccl/restore_span_covering_test.go
Comment (adityamaru) [inline pkg/ccl/backupccl/restore_span_covering_test.go:223]: nit: avoid using TODOSQLCodec

File: pkg/ccl/backupccl/restore_span_covering_test.go
Comment (adityamaru) [inline pkg/ccl/backupccl/restore_span_covering_test.go:68]: nit: Introduce not reintroduce 😋  given they both mean distinct things?

File: pkg/ccl/backupccl/restore_span_covering_test.go
Comment (adityamaru) [inline pkg/ccl/backupccl/restore_span_covering_test.go:61]: So this will drop a span in every incremental layer after `backupWithDroppedSpan`?

File: pkg/ccl/backupccl/restore_span_covering_test.go
Comment (adityamaru) [reply pkg/ccl/backupccl/restore_span_covering_test.go:67]: is this concerning? I haven't looked too deeply but does it warrant understanding why the table ID generator affects the test failing/passing?

File: pkg/ccl/backupccl/restoration_data.go
Comment (adityamaru) [inline pkg/ccl/backupccl/restoration_data.go:143]: nit: span was never reintroduced, given that introduced means that it was new span in between backups.

**Review by adityamaru — CHANGES_REQUESTED** [review]
Overall LGTM, I want to take another pass after this round of comments but mainly of the test files. The test changes are more complicated than the actual code changes. Nice work on making this all work together!

File: pkg/ccl/backupccl/backuppb/backup.proto
Comment (dt) [reply pkg/ccl/backupccl/backuppb/backup.proto:128]: yeah, ditto: it'd be nice to point out that this is a subset, and why we need to track it explicitly (and ideally how it works in upgrades)

File: pkg/ccl/backupccl/backupinfo/manifest_handling.go
Comment (dt) [inline pkg/ccl/backupccl/backupinfo/manifest_handling.go:926]: wouldn't being in _introduced_ spans (as opposed to reintroduced) as of layer X be sufficient to say that we shouldn't be restoring the span from files from layer prior to X? do we really need to specifically track Reintroduced separately?

File: pkg/ccl/backupccl/restore_span_covering.go
Comment (dt) [inline pkg/ccl/backupccl/restore_span_covering.go:109]: nit: I'd call this coveredLater; if anything it is the opposite of covered right now, right?

File: pkg/ccl/backupccl/restore_span_covering_test.go
Comment (dt) [reply pkg/ccl/backupccl/restore_span_covering_test.go:67]: > I feel like the same spans should get covered over a series of backups.

It does with the i/3 doesn't it: at some point, some span, say, 106, starts getting included, and it keeps getting included for at least 3 backups, before i/3 increases (integer division). In fact, I think it stays covered for much more than 3, since the one at the bottom of j slides out and it slides down within j.

**Review by dt — APPROVED** [review]
I haven't fully grokked the changes in span_covering_test.go but everything else mostly lgtm.

One question inline about whether we really, actually, need to distinguish _reintroduced_ from plan ol introduced, since introduced already implies we fully cover it in the layer in which it was (re)introduced, so can't we just use it?

File: pkg/ccl/backupccl/backupinfo/manifest_handling.go
Comment (msbutler) [reply pkg/ccl/backupccl/backupinfo/manifest_handling.go:926]: not necessarily -- a new index on an old table gets included in `IntroducedSpans`, but we certainly don't want to elide all data from old backups associated with that table. See this [comment](https://github.com/cockroachdb/cockroach/pull/87312#issuecomment-1242545077) for more detail and an example.

You then might wonder, "why not keep a `map[table_index]hlc.timestamp` then?" Well, that leads to problems too, as explained in this [thread.](https://github.com/cockroachdb/cockroach/pull/87312#discussion_r967031256) . I'll lay out the alternatives considered into updated commit message.

Comment (dt) [reply pkg/ccl/backupccl/backupinfo/manifest_handling.go:926]: Isn't there a distinction here between putting just the index span into introduced span versus putting the whole table span in it?

Comment (msbutler) [reply pkg/ccl/backupccl/backupinfo/manifest_handling.go:926]: Aha! The first n spans in backupManifest.IntroducedSpans are the new index spans. On the restore side, i think we can identify the new spans by calling:

`newSpans = filterSpans(latestBackup.Spans, prevBackups[len(prevBackups)-1].Spans)`

, remove them from IntroducedSpans, and get the reIntroducedSpans. 

This means we don't need to modify the backup manifest. Brilliant!

**Review by msbutler — COMMENTED** [review]
RFAL!

**Comment by dt** [comment]
> Specifically, we'll need to divide the IntroducedSpans field into IntroducedSpans (i.e. what foo_idx is) and a new field ReIntroducedSpans which contains all spans that need to be reintroduced due to non-mvcc shenanigans

I'm not sure I follow this. In this case, if foo_idx span is introduced at t5, there's no reason for the _index span_ to be included at any earlier time. If the entire foo _table_ span is introduced at t10, there's no reason to use it from any earlier time. It means that this map by table ID might not be quite the right structure -- we likely need to use a span frontier instead. But I think the core principle is the same: if a span -- any span -- is introduced later, including if it is introduced after being previously covered, i.e. is it is re-introduced, then there's no reason to use its covering from prior layers since we know the last introduction time for that span is total in terms of time.

**Comment by msbutler** [comment]
@dt RFAL with the span frontier impl. One concern, in theory, is span frontier lookup in the makeSimpleImportSpans is log(# of spans to restore). One run of BenchmarkRestoreEntryCover on master vs the PR did not reveal any significant regressions. I could be more rigorous and conduct multiple trials on my gceworker though.

**Comment by msbutler** [comment]
@miretskiy added you to the review in case you had any tips on how I construct the spanFrontier we use to filter out backups  for a given span.

**Review by dt — APPROVED** [review]
I think you can squash the two commits

**Comment by dt** [comment]
It is O(log(introducedSpans)) isn't it? The frontier isn't all spans but just introduced spans right?

**Comment by msbutler** [comment]
Done, but to do that, I added a new util method to the SpanFrontier.

File: pkg/util/span/frontier.go
Comment (msbutler) [inline pkg/util/span/frontier.go:125]: @miretskiy Could I get your blessing on this low impact change here?

File: pkg/ccl/backupccl/backupinfo/manifest_handling.go
Comment (msbutler) [reply pkg/ccl/backupccl/backupinfo/manifest_handling.go]: no longer relevant.

File: pkg/util/span/frontier.go
Comment (miretskiy) [reply pkg/util/span/frontier.go:125]: THis change is :lgtm:

**Comment by msbutler** [comment]
TFTRs!!!!!

bors r=dt

File: pkg/ccl/backupccl/restore_span_covering.go
Comment (adityamaru) [inline pkg/ccl/backupccl/restore_span_covering.go:227]: confirming this is because we want to skip the full?

**Review by adityamaru — APPROVED** [review]

File: pkg/ccl/backupccl/restore_span_covering.go
Comment (msbutler) [reply pkg/ccl/backupccl/restore_span_covering.go:227]: yes. Note that this step doesn't do much because the full backup shouldn't have introduced spans anyway.

**Comment by msbutler** [comment]
bors r=dt, adityamaru

**Comment by msbutler** [comment]
blathers backport 22.2 22.1 21.2

**Pull request merged**
